### ISSUE TITLE
[Proposal] Restructuring PullRequest 

### ISSUE THREAD

**Issue by ryanmrodriguez** [Shopify/shipit-engine#1093]
As an ongoing effort on our part to deliver Review Stacks into Shipit, we have a couple of decisions to make pertaining the structure of our changes. We would like to know your thoughts on the subject matter. 

*What are we trying to do?*
[Deliver a Review Stack feature into Shipit.](https://github.com/Shopify/shipit-engine/issues/960) 

*How are we implementing this?*
We extend the Stack model, and rely on the creation and synchronization of Pull Request objects. This is kicked off with custom handlers.

*What issues are we facing?*
In order to tighten up the lifecycle of Review Stacks, we believe they should be owned by Pull Requests. However, within the current codebase, we feel that a PullRequest doesn't accurately reflect what we are meant to be modeling (a traditional Github Pull Request). Currently the PullRequest class is really only used for a merge queue. So that leads to a few decisions that need to be made before we proceed. 

1) We believe that PullRequest should be split into "MergeRequest" and "PullRequest" to more accurately reflect what each is meant to be. MergeRequest would fill the role of the current PullRequest, and the new PullRequest class would take on modeling a traditional GitHub pull request in the same lightweight fashion as things like Commit, and would be used in our Review Stacks feature. How would you feel about this change? Does it make sense to you? Any other insights?

2) We have read about Merge Queue v2, and we are wondering how this would affect the merge queue v1 feature in Shipit. Is this something you guys can see being completely removed from Shipit in favor of using merge queue v2 (assuming it would be open-sourced)?

**Comment by casperisfine** [comment]
> MergeRequest would fill the role of the current PullRequest, and the new PullRequest class would take on modeling a traditional GitHub pull request in the same lightweight fashion as things like Commit, and would be used in our Review Stacks feature. How would you feel about this change? Does it make sense to you? Any other insights?

Yeah I'm fine with that. That's actually a design decision I've come to regret a while ago. So I'm very much ok for at least a rename of `PullRequest` into `MergeRequest`. Further refactoring would be on a case by case basis. And I'd be fine with actual PullRequest tracking in the Commit fashion, as long as it's optional.

> We have read about Merge Queue v2, and we are wondering how this would affect the merge queue v1 feature in Shipit

What's been publicly presented as "Merge Queue v2" is an entirely distinct project from `shipit-engine`. It might be open sourced at some point, but if it is it won't be part of `shipit-engine`. So I'd say don't be concerned with that.

**Comment by benlangfeld** [comment]
> > We have read about Merge Queue v2, and we are wondering how this would affect the merge queue v1 feature in Shipit
> 
> 
> 
> What's been publicly presented as "Merge Queue v2" is an entirely distinct project from `shipit-engine`. It might be open sourced at some point, but if it is it won't be part of `shipit-engine`. So I'd say don't be concerned with that.

Are you able to confirm that the intent is not for that project to supersede shipit-engine? Are they intended to be complementary or competing tools? Are there any thoughts of removing the existing merge queue functionality from shipit-engine in favour of that new project?

**Comment by casperisfine** [comment]
> Are you able to confirm that the intent is not for that project to supersede shipit-engine?

I don't work on the project and have only a superficial understanding of the project, so.... But the internal code-name is `shipit-next` so yeah, it kinda is a `shipit-engine` alternative. However it's much more centered around kubernetes & such so it's not like we're ever going to push it in this repo to replace the current `shipit-engine`. But yes, we're not as actively developing `shipit-engine`, I'm however committed to maintaining it.

> Are there any thoughts of removing the existing merge queue functionality from shipit-engine in favour of that new project?

Nope, that won't happen. We might stop using it internally, but we have no reason to scrap it.
